# Strategic Challenge 2022

### Bain&Company




### Team Flat101
### Authors:
##### [Adney Costa](https://www.linkedin.com/in/adney-costa-moura-7310531b7/)
##### [Matheus Oliveira](https://www.linkedin.com/in/matheusoliveira12/)
##### [Nívea de Abreu](https://www.linkedin.com/in/niveaabreu/)
##### [Ricardo Mourão](https://www.linkedin.com/in/ricardo-mour%C3%A3o-rodrigues-filho-0b75441b9/)

# Project Assigment

The main goal of this Assignment is to create a model that predicts the land area destinated to the plantation of main
agricultural crops and to pastures for livestock, in the state of Pará.
This model must be able to assess which are the priority areas (in terms of both region and product type) to invest TNC’s resources. These areas should be the most affected by the expansion of agribusiness, in terms of environmental impact.




### Importing Libraries

In [22]:
import pandas as pd
import numpy as np
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn import preprocessing
from sklearn.preprocessing import MinMaxScaler

In [23]:
from google.colab import files 
uploaded = files.upload()

Saving historical-database.csv to historical-database (2).csv


## Understanding the database

In [139]:
# Importing the CSV File
df = pd.read_csv('historical-database.csv', sep = ';')
df.head(10)

year         city_code product_type product destinated_area
0  01/01/1974  960b4f2c94a2fb2c    temporary  Others            0,00
1  01/01/1974  746cc42bfb8f6b62    temporary  Others            0,00
2  01/01/1974  6cce2bf873870afc    temporary  Others            0,00
3  01/01/1974  4de42e351006a2ae    temporary  Others            0,00
4  01/01/1974  5b6072f8f6d37acc    temporary  Others            0,00
5  01/01/1974  eac302b92f87d9ca    temporary  Others            0,00
6  01/01/1974  ab984377b2dc0284    temporary  Others            0,00
7  01/01/1974  074590cfc1c41ebf    temporary  Others            0,00
8  01/01/1974  0307883b5d063703    temporary  Others            0,00
9  01/01/1974  062ce4c17e84291e    temporary  Others            0,00

In [140]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 73744 entries, 0 to 73743
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   year             73744 non-null  object
 1   city_code        73744 non-null  object
 2   product_type     73744 non-null  object
 3   product          73744 non-null  object
 4   destinated_area  73744 non-null  object
dtypes: object(5)
memory usage: 2.8+ MB


In [141]:
# There is any null values?
df.isnull().sum()

year               0
city_code          0
product_type       0
product            0
destinated_area    0
dtype: int64

In [142]:
# Looking for the unique values in the columns
for column in df.columns:
  print(f'Column "{column}" has {len(df[column].unique())} uniques values')

Column "year" has 44 uniques values
Column "city_code" has 144 uniques values
Column "product_type" has 3 uniques values
Column "product" has 11 uniques values
Column "destinated_area" has 7154 uniques values


#### Casting the variables for ideals dtypes

To get a better understanding of the data, is a good pratice to cast, or least try, to cast the variables types to the ideal dtype of each one

In [143]:
#Convert the column "year" to time-stamp
print(f'The type of column "year" before the casting is: {df.year.dtype}')
df['year'] = pd.to_datetime(df['year'], format="%m/%d/%Y")
print(f'The type of column "year" after the casting is: {df.year.dtype}')
#Getting only the year the column for better understading
df['year'] = pd.DatetimeIndex(df['year']).year
df.head(10)

The type of column "year" before the casting is: object
The type of column "year" after the casting is: datetime64[ns]


year         city_code product_type product destinated_area
0  1974  960b4f2c94a2fb2c    temporary  Others            0,00
1  1974  746cc42bfb8f6b62    temporary  Others            0,00
2  1974  6cce2bf873870afc    temporary  Others            0,00
3  1974  4de42e351006a2ae    temporary  Others            0,00
4  1974  5b6072f8f6d37acc    temporary  Others            0,00
5  1974  eac302b92f87d9ca    temporary  Others            0,00
6  1974  ab984377b2dc0284    temporary  Others            0,00
7  1974  074590cfc1c41ebf    temporary  Others            0,00
8  1974  0307883b5d063703    temporary  Others            0,00
9  1974  062ce4c17e84291e    temporary  Others            0,00

In [144]:
#Convert the column "product_type" to categorical type
print(f'The type of column "product_type" before the casting is: {df.product_type.dtype}')
df['product_type'] = df['product_type'].astype('category')
print(f'The type of column "product_type" after the casting is: {df.product_type.dtype}')
#Getting only the year the column for better understading
df.head(10)

The type of column "product_type" before the casting is: object
The type of column "product_type" after the casting is: category


year         city_code product_type product destinated_area
0  1974  960b4f2c94a2fb2c    temporary  Others            0,00
1  1974  746cc42bfb8f6b62    temporary  Others            0,00
2  1974  6cce2bf873870afc    temporary  Others            0,00
3  1974  4de42e351006a2ae    temporary  Others            0,00
4  1974  5b6072f8f6d37acc    temporary  Others            0,00
5  1974  eac302b92f87d9ca    temporary  Others            0,00
6  1974  ab984377b2dc0284    temporary  Others            0,00
7  1974  074590cfc1c41ebf    temporary  Others            0,00
8  1974  0307883b5d063703    temporary  Others            0,00
9  1974  062ce4c17e84291e    temporary  Others            0,00

In [145]:
#Convert the column "product" to categorical type
print(f'The type of column "product" before the casting is: {df["product"].dtype}')
df['product'] = df['product'].astype('category')
print(f'The type of column "product" after the casting is: {df["product"].dtype}')
#Getting only the year the column for better understading
df.head(10)

The type of column "product" before the casting is: object
The type of column "product" after the casting is: category


year         city_code product_type product destinated_area
0  1974  960b4f2c94a2fb2c    temporary  Others            0,00
1  1974  746cc42bfb8f6b62    temporary  Others            0,00
2  1974  6cce2bf873870afc    temporary  Others            0,00
3  1974  4de42e351006a2ae    temporary  Others            0,00
4  1974  5b6072f8f6d37acc    temporary  Others            0,00
5  1974  eac302b92f87d9ca    temporary  Others            0,00
6  1974  ab984377b2dc0284    temporary  Others            0,00
7  1974  074590cfc1c41ebf    temporary  Others            0,00
8  1974  0307883b5d063703    temporary  Others            0,00
9  1974  062ce4c17e84291e    temporary  Others            0,00

In [146]:
# Looking further to the 'destinated_area' column we can see that there are some strange values as ',' in it.
df['destinated_area'].value_counts()

,            39035
0,00          5643
50,00          598
10,00          546
100,00         538
             ...  
178306,19        1
168761,17        1
153550,27        1
151607,45        1
165220,87        1
Name: destinated_area, Length: 7154, dtype: int64

In [147]:
#Dropping the conma:
conma_list = list(df[df.destinated_area==","].index)
df.drop(conma_list,inplace=True)
df['destinated_area'].value_counts()

0,00         5643
50,00         598
10,00         546
100,00        538
300,00        442
             ... 
178306,19       1
168761,17       1
153550,27       1
151607,45       1
165220,87       1
Name: destinated_area, Length: 7153, dtype: int64

In [148]:
# Replace the conma separator for '.'
df['destinated_area'] = df['destinated_area'].apply(lambda x: float(x.split()[0].replace(',', '.')))
df

year         city_code product_type    product  destinated_area
0      1974  960b4f2c94a2fb2c    temporary     Others             0.00
1      1974  746cc42bfb8f6b62    temporary     Others             0.00
2      1974  6cce2bf873870afc    temporary     Others             0.00
3      1974  4de42e351006a2ae    temporary     Others             0.00
4      1974  5b6072f8f6d37acc    temporary     Others             0.00
...     ...               ...          ...        ...              ...
73739  2013  74d75dde6dc4a5ec      pasture  Livestock        150866.56
73740  2014  74d75dde6dc4a5ec      pasture  Livestock        152446.40
73741  2015  74d75dde6dc4a5ec      pasture  Livestock        154681.02
73742  2016  74d75dde6dc4a5ec      pasture  Livestock        164706.28
73743  2017  74d75dde6dc4a5ec      pasture  Livestock        165220.87

[34709 rows x 5 columns]

### Outliers

Existem 39035 valores com vírgulas desnecessários para os cálculos

In [15]:
#df['destinated_area'].value_counts()
lista_virgulas = list(df[df.destinated_area==","].index)
df[df.destinated_area==","]

year         city_code product_type   product destinated_area
6292   01/01/1974  960b4f2c94a2fb2c    temporary      Rice               ,
6293   01/01/1974  960b4f2c94a2fb2c    temporary     Beans               ,
6294   01/01/1974  960b4f2c94a2fb2c    temporary   Cassava               ,
6295   01/01/1974  960b4f2c94a2fb2c    temporary      Corn               ,
6296   01/01/1974  960b4f2c94a2fb2c    temporary       Soy               ,
...           ...               ...          ...       ...             ...
68983  01/01/2001  1bd7145ed1da044a    permanent  Palm oil               ,
68988  01/01/2001  f1fd357c5adaa8dc    permanent  Palm oil               ,
68989  01/01/2001  8a4ab531676691da    permanent  Palm oil               ,
68990  01/01/2001  c45d8ddba643c0d7    permanent      Açaí               ,
68991  01/01/2001  c45d8ddba643c0d7    permanent     Cocoa               ,

[39035 rows x 5 columns]

In [16]:
# removendo valores com ,
df = df.drop(lista_virgulas)
# trocando virgula por ponto
df['destinated_area'] = df['destinated_area'].apply(lambda x: float(x.split()[0].replace(',', '.')))
df

year         city_code product_type    product  destinated_area
0      01/01/1974  960b4f2c94a2fb2c    temporary     Others             0.00
1      01/01/1974  746cc42bfb8f6b62    temporary     Others             0.00
2      01/01/1974  6cce2bf873870afc    temporary     Others             0.00
3      01/01/1974  4de42e351006a2ae    temporary     Others             0.00
4      01/01/1974  5b6072f8f6d37acc    temporary     Others             0.00
...           ...               ...          ...        ...              ...
73739  01/01/2013  74d75dde6dc4a5ec      pasture  Livestock        150866.56
73740  01/01/2014  74d75dde6dc4a5ec      pasture  Livestock        152446.40
73741  01/01/2015  74d75dde6dc4a5ec      pasture  Livestock        154681.02
73742  01/01/2016  74d75dde6dc4a5ec      pasture  Livestock        164706.28
73743  01/01/2017  74d75dde6dc4a5ec      pasture  Livestock        165220.87

[34709 rows x 5 columns]